### PR TITLE
Search filter page

### PR DESCRIPTION
**Pull request by PenghaiZhang**
#1337 

##### Checklist

- [x] the [contributor license agreement] is signed
- [x] commit message follows [commit guidelines]
- [x] screenshots are included showing significant UI changes

##### Description of change

Create a new setting page on which user can configure the Owner filter, Date modified filter and MIME type filters. 
All the configurations are not saved to the Server until the SAVE button is hit.
A dialog is used to add/edit a MIME type filter.

Also create two REST endpoints to support create, update and delete MIME type filters in batch, and update the Search filter endpoint to use JSON body rather than query params.

This PR also fixes a bug about Search filters being deleted from the database when calling the REST endpoint of saving general Search setting.
Screenshot from 2020-04-09 19-56-51
Screenshot from 2020-04-09 19-57-01

### CODE REVIEW

**Comment by PenghaiZhang** [comment]
Screenshot from 2020-04-15 16-30-06
Screenshot from 2020-04-15 16-31-30

File: Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/SearchFilterSettingsPage.tsx [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+        <List
+          subheader={
+            <ListSubheader disableGutters>
+              {searchFilterStrings.mimetypefiltertitle}
+            </ListSubheader>
+          }
+        >
+          {mimeTypeFilters.map(filter => {
+            return (
+              <ListItem divider={true} key={filter.name}>
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/SearchFilterSettingsPage.tsx:333]: Help wanted here. The key should be unique, but items displayed in this list are possible to have no key (e.g. User creates a new filter and does not click save button). So I am not entirely sure what value should I put here. I don't think name is a good choice.

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/SearchFilterResource.scala [hunk 2]
```diff
@@ -111,58 +105,136 @@ class SearchFilterResource {
       "This endpoint is used to update a search filter. A JSON object representing the updated filter is returned if operation is successful.",
     response = classOf[SearchFilter]
   )
-  def updateSearchFilter(
-      @ApiParam(value = "filter UUID") @PathParam("uuid") uuid: UUID,
-      @ApiParam(value = "filter name", required = true) @QueryParam("name") name: String,
-      @ApiParam(value = "filter types", required = true) @QueryParam("mimeTypes") mimeTypes: java.util.List[
-        String]): Response = {
+  def updateSearchFilter(@ApiParam(value = "filter UUID") @PathParam("uuid") uuid: UUID,
+                         searchFilter: SearchFilter): Response = {
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/SearchFilterResource.scala:109]: Help wanted again...Now we have a JSON body in this endpoint. One problem is if I put some unrecognised properties into the body, the Resteasy will return 500 directly because the marshalling Resteasy does fails. Many thanks to Resteasy, but can I handle this by myself without significant changes ?  I want to return 400 to client rather than 500.

File: Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/MimeTypeList.tsx [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import { getMimeTypeDetail, MimeTypeEntry } from "./SearchFilterSettingsModule";
+import { makeStyles } from "@material-ui/styles";
+import { languageStrings } from "../../../util/langstrings";
+
+const useStyles = makeStyles({
+  item: {
+    padding: 0
+  },
+  list: {
+    maxHeight: 400,
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/MimeTypeList.tsx:18]: Help wanted here. The issue is how to make a MUI list in a MUI dialog scrollable, rather than the dialog itself scrollable.
I had a discussion with Tom regarding this, but did not have any luck with MUI Grid and Flexbox styles so in the end I had to use this old approach. Does anybody have suggestions?

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/SearchFilterResource.scala [hunk 4]
```diff
@@ -111,58 +105,136 @@ class SearchFilterResource {
...
 
+  @PUT
+  @Path("search/filter")
+  @ApiOperation(
+    value = "Update multiple MIME type filters",
+    notes =
+      "This endpoint is used to update multipe MIME type filters. A JSON object representing a collection of updated filters is returned if operation is successful.",
+    response = classOf[ApiBatchOperationResponse],
+    responseContainer = "List"
+  )
```
Comment (nmkae) [inline Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/SearchFilterResource.scala:136]: A quick glance at the API docs page seems to indicate we don't define the HTTP response codes + response data, the annotations are just setting a default response but no indication of 200 versus 202.  This only matters if there's an intention a oEQ client could be generated from the swagger contract or if it's just to render a docs page.

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/SearchFilterResource.scala [hunk 5]
```diff
@@ -111,58 +105,136 @@ class SearchFilterResource {
...
+        case Some(filter) =>
+          searchSettings.getFilters.remove(filter)
+          ApiBatchOperationResponse(filterId, 200, s"MIME type filter $filterId has been deleted.")
+        case None => ApiBatchOperationResponse(filterId, 404, uuidNotFound(filterId))
+      }
+      batchResponses += response
+    })
+
+    updateSettings(searchSettings)
+    Response.status(207).entity(batchResponses).build()
```
Comment (nmkae) [inline Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/SearchFilterResource.scala:219]: Is the 207 being handled by the client?

File: autotest/OldTests/src/test/java/com/tle/webtests/test/webservices/rest/SearchSettingApiTest.java [hunk 6]
```diff
@@ -290,4 +274,117 @@ public void testDeleteSearchFilter() throws Exception {
   private String searchFilterUri(String filterId) {
     return PathUtils.urlPath(context.getBaseUrl(), API_SEARCH_FILTER_PATH, filterId);
   }
+
+  @Test
+  public void testBatchUpdateSearchFilter() throws Exception {
```
Comment (nmkae) [inline autotest/OldTests/src/test/java/com/tle/webtests/test/webservices/rest/SearchSettingApiTest.java:279]: Misc notes about the tests:

- Some of them are covering more than one scenario rather than having a test per scenario
- no usage of soft asserts, the first assert failure will prevent the remainder from being tested.  This is a pain point when you fix one assert only to find another, and another and another....
- repeating assert structures rather than helper methods to do common asserts,  e.g. http response codes with response message

File: Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/MimeTypeFilterEditor.tsx [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+  DialogTitle,
+  TextField
+} from "@material-ui/core";
+import {
+  getMIMETypesFromServer,
+  MimeTypeFilter,
+  MimeTypeEntry,
+  vaidateMimeTypeName
+} from "./SearchFilterSettingsModule";
+import { useEffect, useState } from "react";
```
[the remark this answers is not in the document]
Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/MimeTypeFilterEditor.tsx:16]: I need to add ` "allowSyntheticDefaultImports": true` to tsconfig in order to do this, otherwise I have a TS1259 error

File: Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/MimeTypeFilterEditor.tsx [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+interface MimeTypeFilterEditorProps {
+  open: boolean;
+  onClose: () => void;
+  addOrUpdate: (filter: MimeTypeFilter) => void;
+  mimeTypeFilter?: MimeTypeFilter;
+  updateTemplate: (update: TemplateUpdate) => void;
+}
+
+const MimeTypeFilterEditor = ({
```
[the remark this answers is not in the document]
Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/MimeTypeFilterEditor.tsx:31]: What about `MimeTypeFilterEditingDialog` ?

File: Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/MimeTypeFilterEditor.tsx [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    >
+      <DialogTitle>{DIALOG_TITLE}</DialogTitle>
+      <DialogContent>
+        <TextField
+          margin="dense"
+          label={searchFilterStrings.filternamelabel}
+          value={filterName}
+          required
+          fullWidth
+          helperText={searchFilterStrings.helptext}
```
[the remark this answers is not in the document]
Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/MimeTypeFilterEditor.tsx:111]: No, I will remove it.

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/SearchFilterResource.scala [hunk 10]
```diff
@@ -67,33 +67,27 @@ class SearchFilterResource {
...
   }
 
   @POST
   @Path("search/filter")
   @ApiOperation(
-    value = "Add a search filter",
+    value = "Add a MIME type filter",
     notes =
-      "This endpoint is used to add a search filter. A JSON object representing the new filter is returned if operation is successful.",
+      "This endpoint is used to add a MIME type filter. A JSON object representing the new filter is returned if operation is successful.",
```
[the remark this answers is not in the document]
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/SearchFilterResource.scala:82]: I think it's a good idea, as they're a specific type of filter - there are others.

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/SearchFilterResource.scala (hunk 2 above)
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/SearchFilterResource.scala:109]: Didn't we do some RestEasy customisation recently - possibly even in the marshalling space? Do you recall @PenghaiZhang ?

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/MimeTypeResource.scala [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+import io.swagger.annotations.{Api, ApiOperation}
+import javax.ws.rs.core.Response
+import javax.ws.rs.{GET, Path, Produces}
+import org.jboss.resteasy.annotations.cache.NoCache
+import scala.collection.JavaConverters._
+
+case class MimeTypeDetail(mimeType: String, desc: String)
+
+@NoCache
+@Path("settings/")
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/MimeTypeResource.scala:32]: These really are their own top level resource - indeed they have their own table.

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/MimeTypeResource.scala [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+  def listMimeTypes: Response = {
+    LegacyGuice.mimePrivProvider.checkAuthorised()
+    val mimeEntries =
+      LegacyGuice.mimeTypeService.searchByMimeType(Constants.BLANK, 0, -1).getResults.asScala
+    val mimeTypes = mimeEntries.map(entry => {
+      MimeTypeDetail(entry.getType, entry.getDescription)
+    })
+    Response.ok().entity(mimeTypes).build()
+  }
+}
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/MimeTypeResource.scala:54]: If at a minimum (for now) you add a POST endpoint, then it'll make wrapping it in the test easy. Or, unless you're testing it already with an existing set of mimetypes from an institution import.

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/SearchFilterResource.scala (hunk 5 above)
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/SearchFilterResource.scala:219]: Nope. I've added a comment to this effect @PenghaiZhang .

Also, I think we need to document the approach to handling of batching - i.e. how we're using 207 responses. Either as part of this PR, or as a side PR off `develop` could you please add some into `Dev/docs/restendpoint.md`?

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/SearchFilterResource.scala (hunk 2 above)
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/SearchFilterResource.scala:109]: Ahhh, I think I've found it... `com.tle.web.remoting.resteasy.RestEasyExceptionMapper`

Can we do what you're after there abouts?

File: autotest/OldTests/src/test/java/com/tle/webtests/test/webservices/rest/SearchSettingApiTest.java [hunk 13]
```diff
@@ -290,4 +274,124 @@ public void testDeleteSearchFilter() throws Exception {
...
+    final String uri = searchFilterUri("");
+
+    ArrayNode jsonBody = mapper.createArrayNode();
+    // The batch update endpoint supports creating and updating, so create one filter firstly
```
Comment (edalex-ian) [inline autotest/OldTests/src/test/java/com/tle/webtests/test/webservices/rest/SearchSettingApiTest.java:283]:
```suggestion
    // The batch update endpoint supports creating and updating, so create one filter first
```

File: autotest/OldTests/src/test/java/com/tle/webtests/test/webservices/rest/SearchSettingApiTest.java [hunk 14]
```diff
@@ -290,4 +274,124 @@ public void testDeleteSearchFilter() throws Exception {
...
+        "No Search filters matching UUID: " + newFilterId,
+        filterOne.get(RESPONSE_MESSAGE).asText());
+  }
+
+  @Test
+  public void testRetrieveMimeTypes() throws Exception {
+    final JsonNode initialFilters =
+        getEntity(context.getBaseUrl() + "api/settings/mimetype", getToken());
+    assertEquals(153, initialFilters.size());
+  }
```
Comment (edalex-ian) [inline autotest/OldTests/src/test/java/com/tle/webtests/test/webservices/rest/SearchSettingApiTest.java:396]: This is not specifically related to Search Settings - and so should be in it's own class.

File: autotest/OldTests/src/test/java/com/tle/webtests/test/webservices/rest/SearchSettingApiTest.java (hunk 6 above)
Comment (edalex-ian) [reply autotest/OldTests/src/test/java/com/tle/webtests/test/webservices/rest/SearchSettingApiTest.java:279]: Yes, I agree that this test should be split up, and some helper methods would help - I think we did some of them in the recent work on the tests for the updated Taxonomy endpoints. :thinking: 

As for the soft asserts, I'm not too concerned there and happy for it stay with this at the moment - also matches what's done everywhere else currently.

One further consideration is whether going through the verbosity of using objects to create your JSON is necessary. In this kind of testing - as opposed to writing core integration code - I'm not against straight up strings (possibly read from a resource file - especially with Java's limitations of support for multi-line strings and having to escape double quotes). But I see you're also using the objects you're creating to slightly tweak, so... it depends.

File: Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/SearchFilterSettingsPage.tsx [hunk 15]
```diff
@@ -0,0 +1,402 @@
...
+      <Card className={classes.spacedCards}>
+        <SettingsList subHeading={searchFilterStrings.visibilityconfigtitle}>
+          <SettingsListControl
+            divider
+            primaryText={searchFilterStrings.disableownerfilter}
+            secondaryText={""}
+            control={
+              <SettingsToggleSwitch
+                value={searchSettings.searchingDisableOwnerFilter}
+                setValue={value => {
```
[the remark this answers is not in the document]
Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/SearchFilterSettingsPage.tsx:299]: As you mentioned above, the benefit of having `setOwnerFilter` is to make the attributes of `SettingsToggleSwitch` look nicer, but it's only used once in here. So I removed it.

File: Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/SearchFilterSettingsPage.tsx [hunk 16]
```diff
@@ -0,0 +1,402 @@
...
+
+    // Update the collection of changed MIME type filters
+    // if 'filter' comes from the editing of 'selectedMimeTypeFilter' and 'selectedMimeTypeFilter'
+    // exists in 'changedMimeTypeFilters', then replace 'selectedMimeTypeFilter' with 'filter';
+    // otherwise add 'filter' to 'changedMimeTypeFilters'
+    if (
+      selectedMimeTypeFilter &&
+      changedMimeTypeFilters.indexOf(selectedMimeTypeFilter) > -1
+    ) {
+      const index = changedMimeTypeFilters.indexOf(selectedMimeTypeFilter);
```
[the remark this answers is not in the document]
Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/SearchFilterSettingsPage.tsx:190]: Yes, `indexOf` does strict equality comparison, I think.

File: Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/SearchFilterSettingsPage.tsx [hunk 17]
```diff
@@ -0,0 +1,402 @@
...
+        )
+        .then(
+          (): Promise<any> =>
+            // Filters stored in 'deletedMimeTypeFilters' always have an id
+            deletedMimeTypeFilters.length
+              ? batchDelete(
+                  deletedMimeTypeFilters.map(filter => filter.id!)
+                ).catch(error => handleError(error))
+              : Promise.resolve()
+        )
```
[the remark this answers is not in the document]
Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/SearchFilterSettingsPage.tsx:271]: As discussed, let me leave this repeating pattern at the moment.

File: Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/SearchFilterSettingsPage.tsx [hunk 18]
```diff
@@ -0,0 +1,402 @@
...
+
+  // changedMimeTypeFilters contains filters to be added or edited on the Server
+  const [changedMimeTypeFilters, setChangedMimeTypeFilters] = useState<
+    MimeTypeFilter[]
+  >([]);
+
+  // deletedMimeTypeFilters contains filters to be deleted from the Server
+  const [deletedMimeTypeFilters, setDeletedMimeTypeFilters] = useState<
+    MimeTypeFilter[]
+  >([]);
```
[the remark this answers is not in the document]
Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/SearchFilterSettingsPage.tsx:95]: As discussed, let me keep this for now.

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/SearchFilterResource.scala (hunk 2 above)
Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/settings/SearchFilterResource.scala:109]: As I recall,  when we were fixing the [resource-centre log issue](https://github.com/openequella/openEQUELLA/issues/1071), we came across this file but we did not make any changes. But in this PR, I add one change to handle unmarshalling failure

**Comment by PenghaiZhang** [comment]
Resteasy returned 400 if I put some unknown properties in the JSON body.
image

**Comment by PenghaiZhang** [comment]
A dialog is displayed to show all error messaged included in the 207 response.
image

**Comment by PenghaiZhang** [comment]
I have not got a POST endpoint for the MIME type as adding a MIME type seems be a bit complicated.

File: Source/Plugins/Core/com.equella.core/js/tsrc/api/BatchOperationResponse.ts [hunk 19]
```diff
@@ -0,0 +1,16 @@
+export interface BatchOperationResponse {
+  id: string;
+  status: number;
+  message: string;
+}
+
+/**
+ * Group responses that have non 2xx codes and return their messages
```
Comment (nmkae) [inline Source/Plugins/Core/com.equella.core/js/tsrc/api/BatchOperationResponse.ts:8]: JS doc mismatch with the `filter` which is not non-200, it's anything above 400.

Should this be the 2xx range?  Or are 3xx considered OK as well?

File: Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/MimeTypeFilterEditingDialog.tsx [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+} from "./SearchFilterSettingsModule";
+import { useEffect, useState } from "react";
+import { commonString } from "../../../util/commonstrings";
+import MimeTypeList from "./MimeTypeList";
+import { languageStrings } from "../../../util/langstrings";
+import { templateError, TemplateUpdate } from "../../../mainui/Template";
+import { fromAxiosError } from "../../../api/errors";
+import { addElement, deleteElement } from "../../../util/ImmutableArrayUtil";
+
+interface MimeTypeFilterEditorProps {
```
Comment (nmkae) [inline Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/MimeTypeFilterEditingDialog.tsx:24]: Interface no longer matches component name

File: Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/MimeTypeFilterEditingDialog.tsx [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+    addOrUpdate({
+      id: mimeTypeFilter?.id,
+      name: filterName,
+      mimeTypes: selectedMimeTypes
+    });
+    onClose();
+  };
+
+  const isNameValid = vaidateMimeTypeName(filterName);
+  const DIALOG_TITLE = mimeTypeFilter
```
Comment (nmkae) [inline Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/MimeTypeFilterEditingDialog.tsx:109]: Not sure this needs to be uppercase (or exist at all)

File: Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/MimeTypeList.tsx [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+    overflow: "auto"
+  },
+  listTitle: {
+    marginTop: 20
+  }
+});
+
+interface MimeTypeFilterListProps {
```
Comment (nmkae) [inline Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/MimeTypeList.tsx:26]: JS doc missing.

File: Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/SearchFilterSettingsModule.ts [hunk 23]
```diff
@@ -0,0 +1,48 @@
+import Axios from "axios";
+import { encodeQuery } from "../../../util/encodequery";
+import {
+  BatchOperationResponse,
+  groupErrorMessages
+} from "../../../api/BatchOperationResponse";
+
+export interface MimeTypeFilter {
```
Comment (nmkae) [inline Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/SearchFilterSettingsModule.ts:8]: JS doc missing

File: Source/Plugins/Core/com.equella.core/js/tsrc/util/ImmutableArrayUtil.ts [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+  replaceElement: T
+): Array<T> {
+  const index = objects.indexOf(element);
+  if (index < 0) {
+    return addElement(objects, replaceElement);
+  }
+  const newObjects = [...objects];
+  newObjects[index] = replaceElement;
+  return newObjects;
+}
```
Comment (nmkae) [inline Source/Plugins/Core/com.equella.core/js/tsrc/util/ImmutableArrayUtil.ts:24]: I think this may cause some confusion as it's not doing an value equality check for complex objects, it'll add a duplicate object if the reference to the new element is slightly different but the object has exactly the same values

File: Source/Plugins/Core/com.equella.core/js/tsrc/util/ImmutableArrayUtil.ts [hunk 25]
```diff
@@ -0,0 +1,42 @@
+/**
```
Comment (nmkae) [inline Source/Plugins/Core/com.equella.core/js/tsrc/util/ImmutableArrayUtil.ts:1]: I'd be weary of stating these are immutable arrays, I would a immutable array utility to deep clone arrays, this is just shallow clones?

If this is intended to be a commonly used utility then it should have tests to sanity check variations

File: Source/Plugins/Core/com.equella.core/js/tsrc/util/ImmutableArrayUtil.ts [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+ * If an element is found in the provided array, then remove this element,
+ * and return a new array; otherwise return the provided array.
+ */
+export function deleteElement<T>(
+  objects: Array<T>,
+  element: T,
+  deleteCount: number
+): Array<T> {
+  const index = objects.indexOf(element);
```
Comment (nmkae) [inline Source/Plugins/Core/com.equella.core/js/tsrc/util/ImmutableArrayUtil.ts:35]: Again this checking references rather than a value equality check, it may be misleading

File: Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/MimeTypeFilterEditingDialog.tsx [hunk 27]
```diff
@@ -0,0 +1,157 @@
...
+  /**
+   * If true, the dialog will be shown.
+   */
+  open: boolean;
+  /**
+   * Fired when the dialog is closed.
+   */
+  onClose: () => void;
+  /**
+   * Fired when clicking the ADD or OK button
```
Comment (SammyIsConfused) [inline Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/MimeTypeFilterEditingDialog.tsx:32]: These comments show up in storybook as documentation, so its best to be consistent with punctuation - ensure each of these interface and component comments ends with a full stop.

File: Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/MimeTypeList.tsx [hunk 28]
```diff
@@ -0,0 +1,79 @@
...
+  /**
+   * A list of MIME types to be displayed
+   */
+  entries: MimeTypeEntry[];
+  /**
+   * Fired when the value of each Checkbox is changed
+   */
+  onChange(check: boolean, mimeType: string): void;
+
```
Comment (SammyIsConfused) [inline Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/MimeTypeList.tsx:35]: Unnecessary whitespace.

File: Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/SearchFilterSettingsPage.tsx [hunk 29]
```diff
@@ -0,0 +1,389 @@
...
+  Card,
+  CardActions,
+  ListItem,
+  ListItemSecondaryAction,
+  ListItemText,
+  ListSubheader,
+  IconButton,
+  List
+} from "@material-ui/core";
+import { makeStyles } from "@material-ui/styles";
```
Comment (SammyIsConfused) [inline Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/SearchFilterSettingsPage.tsx:19]: Import this from @material-ui/core instead.

File: Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/SearchFilterSettingsPage.tsx [hunk 30]
```diff
@@ -0,0 +1,389 @@
...
+import EditIcon from "@material-ui/icons/Edit";
+import DeleteIcon from "@material-ui/icons/Delete";
+import AddCircleIcon from "@material-ui/icons/AddCircle";
```
Comment (SammyIsConfused) [inline Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/SearchFilterSettingsPage.tsx:22]: All these icon imports could go in a single import.

File: Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/SearchFilterSettingsPage.tsx [hunk 31]
```diff
@@ -0,0 +1,389 @@
...
+  const [messageDialogMessages, setMessageDialogMessages] = useState<string[]>(
+    []
+  );
+
+  useEffect(() => {
+    updateTemplate(tp => ({
+      ...templateDefaults(searchFilterStrings.name)(tp),
+      backRoute: routes.Settings.to
+    }));
+  }, []);
```
Comment (SammyIsConfused) [inline Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/SearchFilterSettingsPage.tsx:122]: Rather than have multiple dependency-less useEffect hooks, consider just consolidating this one and the one below into one hook.

File: Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/MimeTypeFilterEditingDialog.tsx (hunk 27 above)
Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/js/tsrc/settings/Search/searchfilter/MimeTypeFilterEditingDialog.tsx:32]: Thank you Sam for reminding me of this issue!
